### ISSUE TITLE
Attack successfully all the time?

### ISSUE THREAD

**Issue by machengcheng2016** [BorealisAI/advertorch#29]
Hello!
I learn how to use advertorch from this site, https://github.com/BorealisAI/advertorch/blob/master/advertorch_examples/tutorial_attack_defense_bpda_mnist.ipynb. 
I wonder how to know if the current attack works successfully? 'Successfully' means that adversarial samples can be generated under the given constraint, and it is normal that no adversarial samples generated in some robust situation. However, I can't see any cues to judge whether the current attack works.
Correspondingly, another similar toolbox called Foolbox will prompt a *None* warning when the attack fails.

**Comment by gwding** [comment]
This is a design choice that we make. Because there could many meanings of success, e.g. changing the label, changing the label with high confidence, increasing the loss to a certain level and etc. Therefore, we choose to let the user do the judgement by themselves. When success is defined by "changing the label", this could be done by something like
```
from advertorch.utils import predict_from_logits
label == predict_from_logits(adversary.predict(inputs))
```

**Comment by machengcheng2016** [comment]
Got it! In that case, I'd like to define "attack success" by "changing the label", just same as the common definition.
Moreover, I find there are many differences in the attack codes among advertorch, Foolbox, and cleverhans. For example, there is a 'return_early' mechanism implemented in 'BIM' attack in Foolbox, but not in advertorch and cleverhans, which may causes a more subtle perturbation. Besides, 'FGSM' attack in Foolbox gradually increase the *eps* parameter until input is successfully attacked, while *eps* is fixed as 0.3 in advertorch and cleverhans by default, which may leads to a failure to attack.
Here is a question which confuses me for a long time. When I try to do some attack and defense experiments, which toolbox on earth should I choose, and which toolbox is more standard? May you give me some guidance, please?

**Comment by gwding** [comment]
In my opinion (which is probably biased), 
`foolbox` has some tweaks to make attacks more adaptive and easier to be successful.
`advertorch` and `cleverhans` are more "loyal" to the original methods from the papers, and they also support batch attacks, which is needed for adversarial training due to the requirement of generation speed.

My designing goal of advertorch is the following as written in the technical report
```
Compared to existing adversarial robustness related toolboxes (Papernot et al.,
2016a; Rauber et al., 2017), advertorch aims for
1. simple and consistent APIs for attacks and defenses;
2. concise reference implementations, utilizing the dynamic computational graphs in PyTorch;
and
3. fast executions with GPU-powered PyTorch implementations, which are important for “attackin-the-loop” algorithms, e.g. adversarial training.
```

**Comment by machengcheng2016** [comment]
Exactly. The newest beta version, Foolbox 2.0, was released several months ago, which is able to do batch attack. However, I'd rather use advertorch because of the "loyalty" to the original papers :D
